### ISSUE TITLE
How should a user delete their own account?

### ISSUE THREAD

**Issue by keithmgould** [graphql-devise/graphql_devise#162]
I am building an authed mutation to allow a user to delete their account. This involves deleting the current_resource (user).

Once I do this, the devise token auth runs into trouble in the callback:

`devise_token_auth-1.1.5/app/controllers/devise_token_auth/concerns/set_user_by_token.rb:97:in update_auth_header`

with: `Couldn't find User with 'id'=1`

Wondering what graphql_devise would like me to do in this case? How do I signal to devise token auth not to include user data in the auth header?

I looked in graphql_devise's logout mutation but that does not really have what i need since it assumes the current_resource is still around. for example adding `remove_resource` in my mutation does not solve.

Thank you!

**Comment by 00dav00** [comment]
Hi @keithmgould, have you consider executing the deletion async?
In that way you can :
- Mark the user as not able to log in anymore
- Enqueue a job to delete it
- Logout the user from FE
- Delete async in a separate task

Do you have a stack trace of this problem? Would like to understand a bit more on what is happening here.

**Comment by keithmgould** [comment]
Hi hi,

Async is a good idea, though in a sense that introduces a race condition: if the worker (somehow) works fast enough we could still get the user not found error. I'm super open to it, but would (if possible) handle this situation sync and see what it takes.

The stack trace (which shows the double render error, but thats not the core issue- the core issue is that devise token auth tries to run update_auth_header in a callback, I believe once GraphQL is done):

```
  ↳ app/services/user_destroyer_service.rb:4:in `destroy'
  User Load (0.4ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]
Couldn't find User with 'id'=1
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/relation/finder_methods.rb:339:in `raise_record_not_found_exception!'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/relation/finder_methods.rb:451:in `find_one'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/relation/finder_methods.rb:433:in `find_with_ids'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/relation/finder_methods.rb:69:in `find'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/querying.rb:21:in `find'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/core.rb:160:in `find'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/persistence.rb:809:in `block in reload'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/scoping/default.rb:34:in `block in unscoped'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/relation.rb:407:in `block in scoping'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/relation.rb:784:in `_scoping'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/relation.rb:407:in `scoping'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/scoping/default.rb:34:in `unscoped'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/persistence.rb:809:in `reload'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/attribute_methods/dirty.rb:31:in `reload'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/associations.rb:255:in `reload'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/autosave_association.rb:230:in `reload'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/devise_token_auth-1.1.5/app/controllers/devise_token_auth/concerns/set_user_by_token.rb:97:in `update_auth_header'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.5/lib/active_support/callbacks.rb:428:in `block in make_lambda'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.5/lib/active_support/callbacks.rb:249:in `block in halting'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.5/lib/active_support/callbacks.rb:517:in `block in invoke_after'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.5/lib/active_support/callbacks.rb:517:in `each'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.5/lib/active_support/callbacks.rb:517:in `invoke_after'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.5/lib/active_support/callbacks.rb:136:in `run_callbacks'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/abstract_controller/callbacks.rb:41:in `process_action'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_controller/metal/rescue.rb:22:in `process_action'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_controller/metal/instrumentation.rb:33:in `block in process_action'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.5/lib/active_support/notifications.rb:180:in `block in instrument'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.5/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.5/lib/active_support/notifications.rb:180:in `instrument'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_controller/metal/instrumentation.rb:32:in `process_action'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_controller/metal/params_wrapper.rb:245:in `process_action'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/railties/controller_runtime.rb:27:in `process_action'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/abstract_controller/base.rb:136:in `process'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionview-6.0.3.5/lib/action_view/rendering.rb:39:in `process'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_controller/metal.rb:190:in `dispatch'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_controller/metal.rb:254:in `dispatch'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/routing/route_set.rb:50:in `dispatch'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/routing/route_set.rb:33:in `serve'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/journey/router.rb:49:in `block in serve'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/journey/router.rb:32:in `each'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/journey/router.rb:32:in `serve'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/routing/route_set.rb:834:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/warden-1.2.9/lib/warden/manager.rb:36:in `block in call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/warden-1.2.9/lib/warden/manager.rb:34:in `catch'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/warden-1.2.9/lib/warden/manager.rb:34:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/rack-2.2.3/lib/rack/tempfile_reaper.rb:15:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/rack-2.2.3/lib/rack/etag.rb:27:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/rack-2.2.3/lib/rack/conditional_get.rb:40:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/rack-2.2.3/lib/rack/head.rb:12:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/http/content_security_policy.rb:18:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:266:in `context'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:260:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/middleware/cookies.rb:648:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.5/lib/active_record/migration.rb:567:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/middleware/callbacks.rb:27:in `block in call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.5/lib/active_support/callbacks.rb:101:in `run_callbacks'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/middleware/callbacks.rb:26:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/middleware/executor.rb:14:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/middleware/actionable_exceptions.rb:18:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/rollbar-3.1.2/lib/rollbar/middleware/rails/rollbar.rb:25:in `block in call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/rollbar-3.1.2/lib/rollbar.rb:145:in `scoped'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/rollbar-3.1.2/lib/rollbar/middleware/rails/rollbar.rb:22:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/middleware/debug_exceptions.rb:32:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/rollbar-3.1.2/lib/rollbar/middleware/rails/show_exceptions.rb:22:in `call_with_rollbar'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/web-console-4.1.0/lib/web_console/middleware.rb:132:in `call_app'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/web-console-4.1.0/lib/web_console/middleware.rb:28:in `block in call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/web-console-4.1.0/lib/web_console/middleware.rb:17:in `catch'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/web-console-4.1.0/lib/web_console/middleware.rb:17:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/railties-6.0.3.5/lib/rails/rack/logger.rb:37:in `call_app'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/railties-6.0.3.5/lib/rails/rack/logger.rb:26:in `block in call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.5/lib/active_support/tagged_logging.rb:80:in `block in tagged'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.5/lib/active_support/tagged_logging.rb:28:in `tagged'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.5/lib/active_support/tagged_logging.rb:80:in `tagged'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/railties-6.0.3.5/lib/rails/rack/logger.rb:26:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/sprockets-rails-3.2.2/lib/sprockets/rails/quiet_assets.rb:13:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/middleware/request_id.rb:27:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/rack-2.2.3/lib/rack/method_override.rb:24:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/rack-2.2.3/lib/rack/runtime.rb:22:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.5/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/middleware/executor.rb:14:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/middleware/static.rb:126:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/rack-2.2.3/lib/rack/sendfile.rb:110:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.5/lib/action_dispatch/middleware/host_authorization.rb:82:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/webpacker-5.2.1/lib/webpacker/dev_server_proxy.rb:25:in `perform_request'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/rack-proxy-0.6.5/lib/rack/proxy.rb:57:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/railties-6.0.3.5/lib/rails/engine.rb:527:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/puma-4.3.7/lib/puma/configuration.rb:228:in `call'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/puma-4.3.7/lib/puma/server.rb:713:in `handle_request'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/puma-4.3.7/lib/puma/server.rb:472:in `process_client'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/puma-4.3.7/lib/puma/server.rb:328:in `block in run'
/Users/keithgould/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/puma-4.3.7/lib/puma/thread_pool.rb:134:in `block in spawn_thread'
Completed 500 Internal Server Error in 677ms (Views: 0.2ms | ActiveRecord: 142.9ms | Allocations: 126534)



AbstractController::DoubleRenderError (Render and/or redirect were called multiple times in this action. Please note that you may only call render OR redirect, and at most once per action. Also note that neither redirect nor render terminate execution of the action, so if you want to exit an action after redirecting, you need to do something like "redirect_to(...) and return".):

app/controllers/concerns/exception_handler.rb:21:in `block (2 levels) in <module:ExceptionHandler>'
```

**Comment by mcelicalderon** [comment]
Hey @keithmgould! Are you refreshing tokens in every request?

**Comment by keithmgould** [comment]
Hey @mcelicalderon I don't think so? But not sure. I'm new to both graphql-devise and devise_auth_token, and I don't think I explicitly set that up. I sort of just followed main instructions for graphql-devise and it worked (which is cool, but leaves me a bit uneducated)

**Comment by mcelicalderon** [comment]
Mmm yes you are. And I think I found the problem https://github.com/lynndylanhurley/devise_token_auth/blob/0c369d6338a362c1e888378b46e8251d8e5de947/docs/usage/controller_methods.md#methods
<img width="1078" alt="image" src="https://user-images.githubusercontent.com/3926414/112049543-c62bbe00-8b1d-11eb-8570-00c47b1fb41f.png">
From DTA docs, looks like this is a special scenario they do handle. From a quick look at [DTA's code](https://github.com/lynndylanhurley/devise_token_auth/blob/05696c7be0a71a4b668cdcaf87e7a304edeb4454/app/controllers/devise_token_auth/concerns/set_user_by_token.rb#L103) you might be able to prevent this by setting the controller instance variable `@used_auth_by_token` to false on that specific mutation. We need to add a better way to do this on our gem, but for now remember you do have access to the controller on the mutation's controller context `context[:controller]`

**Comment by mcelicalderon** [comment]
ok, so you might have run our generator on a version where we didn't default to not refreshing tokens in each request. Take a look at [this](https://github.com/graphql-devise/graphql_devise#devise-token-auth-initializer). Setting `change_headers_on_each_request` to false on `devise_token_auth.rb` initializer will fix this, but be aware that it will change the behavior for a token to be valid for unlimited requests until it expires. More on this [here](https://devise-token-auth.gitbook.io/devise-token-auth/config/initialization)

**Comment by keithmgould** [comment]
Already had that set to false, so I don't believe thats the issue:

![image](https://user-images.githubusercontent.com/40254/112051032-d47ed780-8b27-11eb-9437-15ffb14c0961.png)

And from Gemfile.lock in case it helps:

![image](https://user-images.githubusercontent.com/40254/112051110-eceef200-8b27-11eb-9555-5bc7977e0d32.png)

I'll try setting @used_auth_by_token to false and see.

Thank you for help so far!

**Comment by mcelicalderon** [comment]
Actually, I missread DTA's code 😆 it's the opposite, so setting that variable won't help. But actually found a good fix, you need to do what we already do in the logout mutation in your delete account mutation. Call this method as we do [here](https://github.com/graphql-devise/graphql_devise/blob/b9ab09e8bc2dd0b1a3e359c0c1f409764f16f248/lib/graphql_devise/mutations/logout.rb#L11). That will prevent the resource reload and token generation attempt.

**Comment by keithmgould** [comment]
Hey @mcelicalderon I tried that before opening ticket (mentioned above) and it did not help. But I'll double check it was the same error (and not a method not found error etc...)

**Comment by mcelicalderon** [comment]
Not sure how you are getting passed [this line](https://github.com/lynndylanhurley/devise_token_auth/blob/v1.1.5/app/controllers/devise_token_auth/concerns/set_user_by_token.rb#L89) after calling that method. I see in your backtrace something about a double render. You need to make sure this method is called before mutation execution ends and control returns to the rails controller.

**Comment by keithmgould** [comment]
Just stepped through it. Getting past that line because the instance variable `@resource` exists and is present, but is stale. See me recreate raise in debuguger:

```
(byebug) @resource
#<User id: 1, provider: "email", uid: "keith@gmail.com", allow_password_change: false, uuid: "f227b618-4745-4b0a-a78b-c59f7e766609", nickname: "foo", email: "keith@gmail.com", created_at: "2021-03-22 20:15:30", updated_at: "2021-03-22 20:16:18">
(byebug) @resource.reload
  User Load (1.0ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]
  ↳ (byebug):1:in `update_auth_header'
*** ActiveRecord::RecordNotFound Exception: Couldn't find User with 'id'=1

nil
(byebug)
```

**Comment by mcelicalderon** [comment]
That's precisely what `remove_resource` does 😄 

[here](https://github.com/graphql-devise/graphql_devise/blob/b9ab09e8bc2dd0b1a3e359c0c1f409764f16f248/lib/graphql_devise/concerns/controller_methods.rb#L24)

**Comment by keithmgould** [comment]
🙃 losing mind.

**Comment by mcelicalderon** [comment]
I'll take a closer look later, but all that looks alright. Maybe something weird going on in the controller? Something setting the resource again? Check you before and after actions

**Comment by keithmgould** [comment]
So this solved:

```
class InvalidPasswordError < StandardError
  def message
    "Current password incorrect."
  end
end

module Mutations
  module Users
    class Destroy < BaseMutation
      graphql_name "UsersDestroy"

      argument :password, String, required: true
      
      type GraphQL::Types::Boolean

      def resolve(password:)
        raise InvalidPasswordError unless current_user.valid_password?(password)
        UserDestroyerService.destroy(user: current_user)
        
        context[:controller].resource = nil
        context[:controller].client_id = nil
        context[:controller].token = nil
        true
      end
    end
  end
end
```

I'm guessing that `remove_resource` somehow did not reach? when I ran debugger inside of mutation, and tried 

```
controller.resource = nil
```

it failed because the `controller` namespace is of course not available directly from the mutation, we need context. 

I think as final thoughts, it would be neat if this were handled by the graphql_devise gem in some way? 

Most of all though thank you guys for helping!!

Keith

**Comment by mcelicalderon** [comment]
That is really weird, but I'm glad you figured it out. Are you adding this concern to your controller?
```ruby
include GraphqlDevise::Concerns::SetUserByToken
```
if you are still using this one
```ruby
DeviseTokenAuth::Concerns::SetUserByToken
```
that might be the problem as they are no longer interchangeable. Let me know if that's the case, because otherwise I can't understand how 
```ruby
context[:controller].resource = nil
context[:controller].client_id = nil
context[:controller].token = nil
```
works and calling the method that does exactly the same thing doesn't 🤔

**Comment by keithmgould** [comment]
Yes using the new one:

```
class GraphqlController < ApplicationController
  include GraphqlDevise::Concerns::SetUserByToken
  ...
end
```

**Comment by mcelicalderon** [comment]
Very weird. Not many options left. I can only think of
1. Overwriting `remove_resource` in your base mutation?
2. Does your base mutation inherit from ours or you are including the `ControllerMethods` concern?

**Comment by mcelicalderon** [comment]
So that's it, our base mutation [includes a concern](https://github.com/graphql-devise/graphql_devise/blob/b9ab09e8bc2dd0b1a3e359c0c1f409764f16f248/lib/graphql_devise/mutations/base.rb#L8) where the `remove_resource` method is defined. You will need to include this in your delete user mutation for it to be available. Now I wonder why you are not getting an undefined method error when calling it.

**Comment by keithmgould** [comment]
I'm now using your mutation as parent, and the `remove_resource` works just fine.

The answer to the mystery about `remove_resource` not raising a method not found error is that I catch all errors in my main schema as a precaution:

```ruby
class MySchema < GraphQL::Schema
 ...
  rescue_from(ActiveRecord::RecordNotFound) do |err, obj, args, ctx, field|
    Rails.logger.error "Rescuing from ActiveRecord::RecordNotFound and re-raising as GraphQL::ExecutionError"
    raise GraphQL::ExecutionError, "Could not find #{err.model}"
  end

  rescue_from(StandardError) do |err, obj, args, ctx, field|
    Rails.logger.error "Rescuing from standard error and re-raising as GraphQL::ExecutionError"
    raise GraphQL::ExecutionError, err.message
  end
end
```

I'm not sure if this is an anti-pattern, but did not want to ever send down 500s to client. I've now added the log lines shown above to help with debugging.

**Comment by mcelicalderon** [comment]
I've seen that before, so I think it's fine as long as you capture those errors with a tool like [errbit](https://github.com/errbit/errbit) or you are always looking at the logs for errors. One thing I would really advise against is returning the message of the exception to the client as part of the response, that would be a security flaw as you don't want an external user to know what's going on with your server's code. I would just reply with a generic message an catch the actual error and backtrace with Errbit or similar.

Closing this one as I think the original question was resolved 🙌

**Comment by keithmgould** [comment]
> One thing I would really advise against is returning the message of the exception to the client as part of the response

Thats a good point. will fix.

Agreed to close the issue though it might be nice to mention how to handle this (destroying your own account) in readme?

**Comment by mcelicalderon** [comment]
Right, and I was actually thinking we should add a remove_resource mutation to the gem too. DTA does support this action.
